lib/llvmopencl/BarrierTailReplication.cc

```diff
@@ -21,77 +21,87 @@
+#include <llvm/ADT/Twine.h>
+POP_COMPILER_DIAGS
 IGNORE_COMPILER_WARNING("-Wunused-parameter")
-
-#include "pocl.h"
-
-#include "llvm/Transforms/Utils/BasicBlockUtils.h"
-#include "llvm/Transforms/Utils/Cloning.h"
-#include "llvm/IR/InstrTypes.h"
-#include "llvm/IR/Instructions.h"
+#include <llvm/Analysis/LoopInfo.h>
+#include <llvm/IR/InstrTypes.h>
+#include <llvm/IR/Instructions.h>
+#include <llvm/Transforms/Utils/BasicBlockUtils.h>
+#include <llvm/Transforms/Utils/Cloning.h>
+#include <llvm/Transforms/Utils/Local.h>
+#include <llvm/Transforms/Utils/LoopSimplify.h>
 
 #include "Barrier.h"
 #include "BarrierTailReplication.h"
+#include "LLVMUtils.h"
 #include "VariableUniformityAnalysis.h"
 #include "Workgroup.h"
 #include "WorkitemHandlerChooser.h"
 
 POP_COMPILER_DIAGS
 
+//#define DEBUG_BARRIER_REPL
+
+#include <algorithm>
+#include <iostream>
+#include <map>
+#include <set>
+#include <vector>
+
+#define PASS_NAME "barriertails"
+#define PASS_CLASS pocl::BarrierTailReplication
+#define PASS_DESC "Barrier tail replication pass"
+
+namespace pocl {
+
 using namespace llvm;
-using namespace pocl;
 
-//#define DEBUG_BARRIER_REPL
+class BarrierTailReplicationImpl {
 
-static bool block_has_barrier(const BasicBlock *bb);
-  
-namespace {
-  static
-  RegisterPass<BarrierTailReplication> X("barriertails",
-					 "Barrier tail replication pass");
-}
+public:
+  bool runOnFunction(llvm::Function &F);
+  BarrierTailReplicationImpl(llvm::DominatorTree &DT, llvm::LoopInfo &LI)
+      : DT(DT), LI(LI){};
 
-char BarrierTailReplication::ID = 0;
+private:
+  typedef std::set<llvm::BasicBlock *> BasicBlockSet;
+  typedef std::vector<llvm::BasicBlock *> BasicBlockVector;
+  typedef std::map<llvm::Value *, llvm::Value *> ValueValueMap;
 
-void
-BarrierTailReplication::getAnalysisUsage(AnalysisUsage &AU) const
-{
-  AU.addRequired<DominatorTreeWrapperPass>();
-  AU.addPreserved<DominatorTreeWrapperPass>();
-  AU.addRequired<LoopInfoWrapperPass>();
-  AU.addPreserved<LoopInfoWrapperPass>();
+  llvm::DominatorTree &DT;
+  llvm::LoopInfo &LI;
 
-  AU.addPreserved<VariableUniformityAnalysis>();
-  AU.addRequired<WorkitemHandlerChooser>();
-  AU.addPreserved<WorkitemHandlerChooser>();
-}
+  bool ProcessFunction(llvm::Function &F);
+  bool FindBarriersDFS(llvm::BasicBlock *bb, BasicBlockSet &processed_bbs);
+  bool ReplicateJoinedSubgraphs(llvm::BasicBlock *dominator,
+                                llvm::BasicBlock *subgraph_entry,
+                                BasicBlockSet &processed_bbs);
 
-bool
-BarrierTailReplication::runOnFunction(Function &F)
-{
-  if (!isKernelToProcess(F))
-    return false;
-  if (getAnalysis<WorkitemHandlerChooser>().chosenHandler() ==
-      WorkitemHandlerChooser::POCL_WIH_CBS)
-    return false;
+  llvm::BasicBlock *ReplicateSubgraph(llvm::BasicBlock *entry,
+                                      llvm::Function *f);
+  void FindSubgraph(BasicBlockVector &subgraph, llvm::BasicBlock *entry);
```

Comment:
etc. Please check all your C++ commits if there are now variables we could rename in parts we touch anyhow.